scripts/create_gw5000_database.py

```diff
@@ -0,0 +1,142 @@
+import pandas as pd
+import re
+from mendeleev import element
+from collections import Counter
+from tqdm import tqdm
+from dateutil import tz
+from datetime import datetime
+
+def convert_smiles_to_molecular_formula(inchi: str) -> str:
+    ids_slash = [m.start() for m in re.finditer('/', inchi)]
+    start_idx = ids_slash[0] + 1
+    end_idx = ids_slash[1]
+
+    molecular_formula = inchi[start_idx: end_idx]
+    return  molecular_formula
+
+
+def calc_custom_molecular_formula(xyz):
+    atoms = []
+    for line in xyz.split('\n'):
+        str_regex = '[a-zA-Z]+'
+        atom_name = re.findall(str_regex, line)
+
+        atoms.extend(atom_name)
+
+    counter = Counter(atoms)
+
+    atom_names = list(counter.keys())
+    n_atoms = list(counter.values())
+
+    molecular_formula_info = []
+    for atom, n in zip(atom_names, n_atoms):
+        molecular_formula_info.append(atom)
+        molecular_formula_info.append(str(n))
+
+    custom_molecular_formula = ''.join(molecular_formula_info)
+    return custom_molecular_formula
+
+
```

Comment:
For charged molecules, the number of electrons will be different, but this will probably only matter when we're dealing with a database that has geometries for charged molecules.

Reply:
Would you recommend changing the electron calculation to account for charges in this PR itself?

Reply:
So I am able to implement a way to get the formal charge on a molecule using rdkit.
The total time for 62k database grows by about a minute. Now it is 2mins 36 seconds.

I would like to get a few clarifications about my approach:

So I am using the inchi to build the rdkit mol. So for some inchis, I think about 7 of them, rdkit fails to build a mol due to valency conditions like example: 
`[17:33:10] ERROR: Explicit valence for atom # 20 Si, 8, is greater than permitted `

So for such molecules I have currently assigned the formal charge to be: 0

Is this a fair assumption to make? 
If not please let me know any alternate approaches for the problem molecules.

Reply:
In this case all of them have 0. 
But I have written code to account for any molecules which could have a charge and that will be detected with rdkit.

Reply:
H20 has inchi : `InChI=1S/H2O/h1H2`
H20+ has inchi: `InChI=1S/H2O/h1H2/q+1`
H20++ has inchi: `InChI=1S/H2O/h1H2/q+2`

The charge information is embedded in the inchi as far as I understand.
Please observe especially the part close the end of the inchi string. 

Therefore rdkit is able to say that the charge is 0,1 and 2 respectively for these molecules.

Reply:
Okay. Excellent.

Reply:
@vandan-revanur if the charge is already so easy to obtain from the InChI, then why do we need to introduce an additional dependency (rdkit)?

Reply:
If the InChI code already has the charge, then I don't see any need to use rdkit. rdkit is just getting the charge from the InChI code? By removing the dependency on rdkit, we make our script simpler and easier to use (doesn't require installing rdkit) and less prone to errors that would occur if rdkit changes without our knowledge.

Reply:
It does appear to be slightly more complex than I expected. 
For example, C17H32N2.Li has the following Inchi:

`InChI=1S/C17H23N2.Li/c1-7-14-10(3)16(18-12(14)5)9-17-11(4)15(8-2)13(6)19-17;/h9H,7-8H2,1-6H3;/q-1;+1/b16-9-;`

As you see, the charge info shows: `q-1;+1`, so I am not sure which one to choose when we have a choice like this. Could you please weigh in on this?

Reply:
I was under the impression that all of the molecules in the OE62 dataset were neutral. Does the paper say anything about any molecules being charged? Does the original JSON file that we downloaded, say anything about the charge? Does the original JSON file have the ionization energies? Are you able to give a summary of the properties of each molecule, that the original downloaded file has, other than XYZ coordinates?

Reply:
These are the various informations available in the JSONs:

[refcode_csd","canonical_smiles","inchi","number_of_atoms","xyz_pbe_relaxed","energies_occ_pbe","energies_occ_pbe0_vac_tier2","energies_occ_pbe0_water","energies_occ_pbe0_vac_tzvp","energies_occ_pbe0_vac_qzvp","energies_occ_gw_tzvp","energies_occ_gw_qzvp","cbs_occ_gw","energies_unocc_pbe","energies_unocc_pbe0_vac_tier2","energies_unocc_pbe0_water","energies_unocc_pbe0_vac_tzvp","energies_unocc_pbe0_vac_qzvp","energies_unocc_gw_tzvp","energies_unocc_gw_qzvp","cbs_unocc_gw","total_energy_pbe","total_energy_pbe0_vac_tier2","total_energy_pbe0_water","total_energy_pbe0_vac_tzvp","total_energy_pbe0_vac_qzvp","hirshfeld_pbe","hirshfeld_pbe0_vac_tier2","hirshfeld_pbe0_water"]

Reply:
When I run the inchi through rdkit, it gives me a total net charge as : 0.

So my guess is that one of the N has a  -1 charge and the Li has a +1 charge. So total net charge is 0.
This is the structure of the molecule:
![C17H32N2 Li_structure](https://user-images.githubusercontent.com/5222803/233380747-d0a8d153-80be-45de-bdc1-b320dc39ea3d.png)

Does that make sense?

Reply:
[This paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4486400/) says:

> InChI for Cl−, “InChI=1S/ClH/h1H/p-1”

 - `1S ` means "standard InChI" in contrast to non-standard InChI which would just be `1`.
 - `ClH` denotes the ClH molecule, although not in Hill notation, even though the paper says "Empirical formula sublayer: representing composition. The chemical formula is represented according to Hill convention, that is, beginning with carbon atoms, then hydrogens, then all other elements in alphabetical order. This is the only layer prefixed with a single slash, ’/’, without a following character."
 - `/h1H` is "Hydrogens layer. This layer prefixed with ‘/h’ lists the bonds between the atoms in the structure, partitioned into as many as three sublayers. The first sublayer represents all bonds other than those to non-bridging H-atoms, the second sublayer represents bonds of all immobile H-atoms, and the third sublayer provides locations of any mobile H-atoms. This last sublayer represents H-atoms that can be found at more than one location in a compound due to various types of tautomerism. This sublayer identifies the groups of atoms that share one or more mobile hydrogen atoms. In addition to hydrogen atoms, mobile H groups may contain mobile negative charges. These charges are included in the charge layer."
 - `/p-1` means that 1 proton was removed from the ClH, which is why the molecule has a charge of -1. Unfortunately this description alone, which is the best I could get form a brief glance at the paper, does not describe whether the proton is removed from the Cl atom or from the H atom, but the proton is removed from the H atom in this case.
 
The Cl- InChI from that paper is the same here: https://cccbdb.nist.gov/inchix.asp
There we see that HCl- would be `InChI=1S/ClH/h1H/q-1` and H2O- would be `InChI=1S/H2O/h1H2/q-1`.

From this I conclude that `q-1` means a charge of -1, and `q+1` means a charge of +1, and `p-1` means one proton is removed from a hydrogen atom, and `p+1` means that one proton is added (as in the case of H3O+ for which that CCCBDB site gives us `InChI=1S/H2O/h1H2/p+1`). 

The case of C17H23N2.Li with `q-1;+1` probably means that C17H23N2 has a charge of -1 and Li has a charge of +1. You suggested that one of the N's carries the -1 charge, and based on the diagram and symmetry, it might also be reasonable to say that the negative charge is shared between the two seemingly symmetrically equivalent N's (so each N effectively has a charge of -1/2), but where the negative charge is located, is not depicted by the InChI itself.

In terms of how to proceed:
 - Anything with `/q-X/` means a charge of -X
 - Anything with `/q+X/` means a charge of +X
 - Anything with `/q+X;q-Y` or `q-X;q+Y` means a charge of X-Y or Y-X respectively
 - Analogous rules to the above three, can be made whenever we see `/p`

I'm not sure if I'm missing anything beyond the above rules, how long is the code in RDKit for determining net charges?

Reply:
The rdkit code is in C++ and they write python wrappers on top of it. I am not so comfortable messing around with C++ code. 

Also I hit a very interesting Inchi :

`InChI=1S/C19H54N3O3Si7.3Li/c1-19(2,3)32(23-29(13,14)20-26(4,5)6,24-30(15,16)21-27(7,8)9)25-31(17,18)22-28(10,11)12;;;/h1-18H3;;;/q-3;3*+1`

I have no clue how to process that ! 

My suggestion is to use rdkit so that we have something running. We can then create an issue to re-implement this as using string regexes later.

Reply:
I'm just as new to this as you (maybe even newer to it than you because of my lack of familiarity with RDKit and Python, and the fact that I never used the term "InChI" until you mentioned it in these goDatabase discussions!).

However, it looks like C19H54N3O3Si7.3Li is C19H54N3O3Si7 with 3 Li atoms, and `q-3;3*+1` means that C19H54N3O3Si7 has a charge of -3 and `3*+1` means that there's a +1 charge contribution for each Li atom, so `3*` that due to the three Li atoms. 

I totally agree with you about not re-inventing the wheel when it's not productive to do so.
I'm not totally opposed to using RDKit.

But I'm certainly benefiting from this discussion because I'm seeing more and more "corner" cases and the real landscape of what he dataset looks like. This might not feel as much beneficial or productive to you though, so at any point if you want to switch back to using RDKit (knowing what we've discussed now), I will approve it. Hopefully though, this corner case is the last one!

Reply:
I hit another case. 

`InChI=1S/C25H43NO5.Li/c1-24(2,3)21-17-20(23(27)22(18-21)25(4,5)6)19-26-7-9-28-11-13-30-15-16-31-14-12-29-10-8-26;/h17-18,27H,7-16,19H2,1-6H3;/q;+1/p-1`

Note that right after q there is no number but a ; 
I was able to tune my code to handle that.

I now have a running code without using rdkit. 
It is upto us to now figure out whether to continue to use rdkit or use our custom function to calculate the total net charge. 

I vote for rdkit because we still do not know about some possible corner that we have not yet accounted for !
If you choose to go for rdkit, then I will create an issue and mention our custom code there.

Reply:
What did you do in that case? The Li seems to have a charge of +1 but for C25H43NO5 it looks like we have a `q` value of 0 followed`p-1` which would suggest that we have a loss of one hydrogen and a another positive charge (because p+1 means a charge of -1 in the case of Cl-)? Is the formula for this molecule (C25H33NO5.Li) with a charge of +2? This would contradict my understanding that all molecules in the dataset are neutral.

Maybe understanding how charges work in InChI files could be a good question for MMSE or even Chem.SE (possibly even Bioinformatics.SE but probably more suitable for MMSE or Chem.SE).

When you say that you now have a running code without using RDKit, do you mean that there's no more of these strange cases? Everything falls into one of the following categories?

 - /q-X/
 - /q-X,Y*+Z/   
 - /q;+X/
 - /p-X/
 - Some variant of the above forms (e.g. switching the `-` with a `+`), perhaps even mixing the /p-X with a /q+Y?

As for the names of the molecules, as derived from the InChI code, are we sure that we have all chemical formulas right? For example, `InChI=1S/ClH/h1H/p+1` should not be ClH, it should be Cl (with charge = -1) or just Cl-. 

In the case of `C25H43NO5.Li` with `q;+1/p-1`, what is the chemical formula? 

We can use RDKit, but if it weren't for work we did today, I feel like we (and least me) would have a much weaker understanding of what's going on here. 

It's also interesting that there seems to be so many molecules bound to Li vs other elements, which would indicate a bias in the OE62 dataset, which I thought wasn't supposed to have such biases (but I'm just basing this off of the fact that all three of the recent "corner" cases have had a form of a molecule bound to one or more Li atoms, rather than Na or some other element).

Reply:
@ndattani 

1. What did I do: I am calculating the the net charge by summing up the numbers that occur after the /q symbol and that are separated by the ; 
I am assuming a blank is a 0. 
So for C25H43NO5.Li it is q;+1 it would mean 0+1. I haven’t considered the proton part /p for the charge calculation. 

2. When I say fully running code, I mean that it passed through all the ~62k molecules with the inchi falling into one of the categories you listed above. 

3. I agree that this investigation is useful and I can continue it. I can ask a question tomorrow on MMSE regarding charges on inchi string.

Reply:
@vandan-revanur I see issues with your items (1) and (3). I believe that all molecules in this dataset had an electronic charge of 0 when their XYZ coordinates were obtained, so I think your choice of +1 for the electric charge of C25H43NO5.Li was incorrect due to not considering the /p part. Similarly, the chemical formula for `InChI=1S/ClH/h1H/p+1` is just Cl or Cl-, not ClH, so the /p part needs to be considered when choosing the name of the molecule. For cases in which we have a /p part, does RDKit say the chemical formula is the same as what you find between the two slashes after the `1S`? Does RDKit say that `InChI=1S/ClH/h1H/p+1` is Cl-?

Reply:
Small correction to what you said regarding the protonation part. As far as I understand, `p+1` in the inchi means one needs to add a proton (H+) to HCl which results in `HCl.[H+]` which matches with how rdkit reads the inchi: `InChI=1S/ClH/h1H/p+1`

<img width="706" alt="Screenshot 2023-04-24 at 10 39 42" src="https://user-images.githubusercontent.com/5222803/233946071-72837ee1-700a-4d5b-a9d7-d27002ddcb68.png">

Reply:
@vandan-revanur  the InChI I gave is for Cl- not H2Cl+. I provided a link to the original InChI paper, and the CCCBDB website, both of them say that this InChI is for Cl- not H2Cl+. If RDKit is doing it differently, it might be worth asking at MMSE or Chem.SE.

Reply:
Sorry for that confusion. 

Instead of saying: 
> " Similarly, the chemical formula for InChI=1S/ClH/h1H/p+1 is just Cl or Cl-, not ClH, so the /p part needs to be considered when choosing the name of the molecule. "

I can say:
> " Similarly, the chemical formula for InChI=1S/H2O/h1H2/p+1 is for H3O+ not for H2O, so the /p part needs to be considered when choosing the name of the molecule."

So regarding your points (1) and (3) [here](https://github.com/HPQC-LABS/goDatabase/pull/3#discussion_r1174084642):
- (1) You wrote "So for C25H43NO5.Li it is q;+1 it would mean 0+1" but I think everything in the GW5000 database is neutral.
- (2) You wrote "Regarding molecule name I’m picking it straight from the inchi string. So for InChI=1S/C25H43NO5.Li/ I’m picking the name between the two slashes. In this case it would be: # Name = C25H34NO5.Li" but only looking at what's between the slashes would give you H2O instead of H3O+ in the case of `InChI=1S/H2O/h1H2/p+1` and it would give you HCl instead of Cl- in the case of `InChI=1S/ClH/h1H/p-1`.

Reply:
1. I can make some changes to code to incorporate the protonation part for calculating the charge.
2. I feel this would need some more string manipulation which I feel is unnecessary when there is a function ready to use in rdkit. Therefore I would like to use rdkit for this.

For example molecular formula can be extracted in a simple way with rdkit
```
from rdkit.Chem import rdMolDescriptors
inchi_5 = 'InChI=1S/C25H43NO5.Li/c1-24(2,3)21-17-20(23(27)22(18-21)25(4,5)6)19-26-7-9-28-11-13-30-15-16-31-14-12-29-10-8-26;/h17-18,27H,7-16,19H2,1-6H3;/q;+1/p-1'
mol_5 = Chem.MolFromInchi(inchi_5)
print(rdMolDescriptors.CalcMolFormula(mol_5)) # Outputs C25H42LiNO5 (this accounts for the -1 protonation)
```

Reply:
Actually I realised its not that much work to make a custom function for calculating the molecular formula. So now have completely removed rdkit from our script :)

Reply:
Excellent. I was going to suggest, that whatever "string manipulation" code rdkit has, could probably be copied and pasted, so that we take just the ***necessary*** piece of code from rdkit, rather than the entire package (the license file for RDKit is [here](https://github.com/rdkit/rdkit/blob/master/license.txt)).